cilium/CFP-32759-cilium-mtls-via-mtls-agent.md

```diff
@@ -0,0 +1,260 @@
+
+### Components
+
+#### mTLS Agent
+
+The mTLS agent facilitates workload identity based mTLS connections between node agents that are transparent to the workloads making and receiving requests. It runs separately from the Cilium agent as a part of its own daemonset. Whether or not the mTLS agent will be running its own separate network namespace or the root network namespace is an open question. See Should the mTLS Agent be in the root network namespace or a separate network namespace? for more discussion on the implications of the mTLS agents placement. The Datapath example assumes the mTLS agent is in its own network namespace.
+
+To support the pluggable and interoperable mTLS agent model, the proxying node agent must adhere to the following specifications:
+* Support receiving and requesting configuration via the xDS protocol. Not required to implement the entire [xds API](https://github.com/cncf/xds)
+* Support transparent proxying (i.e., forwards the request without the original client or server knowing)
+* Mutually validate workload identity using mTLS 
+* Ability to request workload certificates
+* Support identity-based policy enforcement using the certificates used for mTLS encryption
+* Support the HTTP Connect protocol with specific port and header conventions for agent to agent communication (see [Protocol](#protocol) for information about HTTP Connect tunneling)
+
+One potential implementation of the mTLS agent is the [ztunnel](https://github.com/istio/ztunnel) L4 node proxy. Ztunnel as the mTLS agent would be deployed in its own network namespace and would require the eBPF component implement an L2 redirection mechanism.
+
+#### eBPF Component
+
+The exact function of the transparent redirection provided by the eBPF component is dependent on the networking namespace of the mTLS agent (own network namespace or root network namespace). If mTLS agent is its own network namespace, the eBPF component will redirect relevant packets from workloads and mTLS agents via L2, MAC address and interface rewrites to the node’s mTLS agent. The kernel programs could utilize a BPF map populated by the Cilium agent to determine what traffic should be redirected and to which interface.
+
+```c
+struct local_redirect_key {
+	__u64	id; // Pod IP for workloads and special well-known key value for mTLS agent
+};
+
+struct local_redirect_info {
+	__u16	ifindex; // Interface index of workload or mTLS agent
+	__u8	ifmac[6]; // MAC address of workload or mTLS agent
+};
+```
+
+Map entries for each mTLS-enabled workload would be created for each map on their respective nodes. A map entry would also be created for the mTLS agent present in that node.
+
+If the mTLS agent is in the root network namespace, TPROXY rules could be added directly to the root network namespace by the eBPF component. Existing transparent redirection mechanisms within Cilium (envoy and DNS transparent proxying) could be generalized to support this feature.
```

Comment:
I agree that something _like_ this will be required, but in practice, we'll also need to _carefully_ wargame out this to ensure that this map is not susceptible to cache-poisoning attacks (at least in so far as they are relevant in the Cilium Threat Model).